### ISSUE TITLE
Raspberry Pi and other non-x64 platforms

### ISSUE THREAD

**Issue by luobendewugong** [wekan/wekan-snap#46]
I can install snapd in my raspberrypi, but can't snap install wekan because error: cannot install "wekan": snap not found.

I have try it for many time but failed.

Could you help me. thanks!

details as follow:

pi@raspberrypi:~ $ sudo apt install snapd
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  apparmor libapparmor-perl snap-confine squashfs-tools
  ubuntu-core-launcher
Suggested packages:
  apparmor-profiles apparmor-profiles-extra apparmor-utils
The following NEW packages will be installed:
  apparmor libapparmor-perl snap-confine snapd squashfs-tools
  ubuntu-core-launcher
0 upgraded, 6 newly installed, 0 to remove and 17 not upgraded.
Need to get 4,233 kB/4,955 kB of archives.
After this operation, 27.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://mirrors.zju.edu.cn/raspbian/raspbian stretch/main armhf snapd armhf 2.21-2 [4,233 kB]
Fetched 4,233 kB in 1s (3,213 kB/s)
Preconfiguring packages ...
Selecting previously unselected package libapparmor-perl.
(Reading database ... 124722 files and directories currently installed.)
Preparing to unpack .../0-libapparmor-perl_2.11.0-3+deb9u2_armhf.deb ...
Unpacking libapparmor-perl (2.11.0-3+deb9u2) ...
Selecting previously unselected package apparmor.
Preparing to unpack .../1-apparmor_2.11.0-3+deb9u2_armhf.deb ...
Unpacking apparmor (2.11.0-3+deb9u2) ...
Selecting previously unselected package snap-confine.
Preparing to unpack .../2-snap-confine_2.21-2_armhf.deb ...
Unpacking snap-confine (2.21-2) ...
Selecting previously unselected package ubuntu-core-launcher.
Preparing to unpack .../3-ubuntu-core-launcher_2.21-2_armhf.deb ...
Unpacking ubuntu-core-launcher (2.21-2) ...
Selecting previously unselected package squashfs-tools.
Preparing to unpack .../4-squashfs-tools_1%3a4.3-3+deb9u1_armhf.deb ...
Unpacking squashfs-tools (1:4.3-3+deb9u1) ...
Selecting previously unselected package snapd.
Preparing to unpack .../5-snapd_2.21-2_armhf.deb ...
Unpacking snapd (2.21-2) ...
Setting up squashfs-tools (1:4.3-3+deb9u1) ...
Processing triggers for systemd (232-25+deb9u2) ...
Processing triggers for man-db (2.7.6.1-2) ...
Setting up libapparmor-perl (2.11.0-3+deb9u2) ...
Setting up apparmor (2.11.0-3+deb9u2) ...
Created symlink /etc/systemd/system/sysinit.target.wants/apparmor.service → /lib/systemd/system/apparmor.service.
update-rc.d: warning: start and stop actions are no longer supported; falling back to defaults
diff: /var/lib/apparmor/profiles/.apparmor.md5sums: No such file or directorySetting up snap-confine (2.21-2) ...
Setting up ubuntu-core-launcher (2.21-2) ...
Setting up snapd (2.21-2) ...
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.autoimport.service → /lib/systemd/system/snapd.autoimport.service.
Created symlink /etc/systemd/system/timers.target.wants/snapd.refresh.timer → /lib/systemd/system/snapd.refresh.timer.
Created symlink /etc/systemd/system/sockets.target.wants/snapd.socket → /lib/systemd/system/snapd.socket.
Created symlink /etc/systemd/system/multi-user.target.wants/snapd.service → /lib/systemd/system/snapd.service.
Processing triggers for systemd (232-25+deb9u2) ...
pi@raspberrypi:~ $ sudo snap install wekan
error: cannot install "wekan": snap not found
pi@raspberrypi:~ $ sudo snap install wekan
error: cannot install "wekan": snap not found
pi@raspberrypi:~ $ snap install
error: access denied (try with sudo)
pi@raspberrypi:~ $ snap list
No snaps are installed yet. Try "snap install hello-world".
pi@raspberrypi:~ $ snap find wekan
The search "wekan" returned 0 snaps
pi@raspberrypi:~ $ snap find Wekan

The search "Wekan" returned 0 snaps
pi@raspberrypi:~ $
pi@raspberrypi:~ $ sudo snap install wekan
error: cannot install "wekan": snap not found
pi@raspberrypi:~ $ sudo snap install wekan --channel=beta
error: cannot install "wekan": snap not found
pi@raspberrypi:~ $ sudo snap install wekan --channel=candidate
error: cannot install "wekan": snap not found
pi@raspberrypi:~ $ sudo snap install wekan --channel=edge
error: cannot install "wekan": snap not found
pi@raspberrypi:~ $

**Comment by xet7** [comment]
There is no snap for raspberry pi yet.

Moved to https://github.com/wekan/wekan/issues/1053

Also see:
* https://blog.wekan.team/2018/01/wekan-progress-on-x64-and-arm/index.html
* https://github.com/wekan/wekan.rs

**Comment by kubiko** [comment]
at the moment wekan is only compiled for amd64
I can't say how much sense this would make to run it on pi3
@xet7 perhaps you can tell if there is demand for other architectures like arm64, it should not be big problem to enable build for other architectures...

**Comment by xet7** [comment]
@kubiko 

Yes there is a lot of demand. I would like to add as many other architectures as possible.

Meteor usually compiles, it's just that default build options of Meteor don't have architecture names like arm64 added to build file as supported platforms. I also had some problems with fibers on arm64, it could be that  Rocket.Chat or others have that already solved. I think there is already snap of Rocket.Chat for other architectures.

I would really appreciate your help, pull requests etc in adding support for other architecture.

**Comment by kubiko** [comment]
@xet7 a bit progress here
I have fixed caddy building to support additional architectures
But where I got blocked is meteor, any idea where I can find meteor for other architectures? I see some mentioned that 1.4 should support arm64? But I could not find any way to build meteor bootstrap package from source.
Other blocker is mongodb, only from version 3.3 onwards arm64 is supported.

So before we can move forward:
 - wekan needs to move to mongo 3.3 or newer
 - meteor needs to gain support for arm64

armhf is probably completely out of the question for now

**Comment by kubiko** [comment]
On related note, I also noticed that wekan is now using patched node engine, do we have source somewhere for this? I'd probably make sense to build it from source instead, this will make it easy to support other architectures

**Comment by xet7** [comment]
@kubiko 

Other architectures with Meteor patches at https://github.com/wekan/wekan/issues/1053#issuecomment-393069820

Patched node source is here https://github.com/sandstorm-io/node

Newer [Meteor 1.6.1 changelog](https://docs.meteor.com/changelog.html) says: "The bundled version of MongoDB used by meteor run in development on 64-bit architectures has been updated to 3.4.10.". Wekan is currently at [Meteor 1.6.0.1, there is upgrading issue](https://github.com/wekan/wekan/issues/1522). It need testing does upgrading work. Main problem with newer MongoDB has been that [uploading file produces crash in cfs_gridfs](https://github.com/wekan/wekan/issues/869).

**Comment by xet7** [comment]
Hmm, maybe here is MongoDB crash fix, I have not tested yet https://github.com/reactioncommerce/reaction/pull/2096

**Comment by xet7** [comment]
@kubiko 

It seems architecture build ignore flags do not work and there are also some git errors:

https://launchpadlibrarian.net/400055090/buildlog_snap_ubuntu_xenial_amd64_wekan-devel_BUILDING.txt.gz

https://github.com/wekan/wekan/blob/edge/snapcraft.yaml#L16-L30

**Comment by xet7** [comment]
@kubiko @stolowski 

Now I have a way to build Wekan for Raspberry Pi, arm64 servers and other architectures, anywhere where is Node 8.16.0 and MongoDB 3.6.x or newer available:
https://github.com/wekan/wekan/wiki/Raspberry-Pi

Wekan Meteor 1.8.1 version is at https://github.com/wekan/wekan master branch. I have tried to make it working at snap core18, but I can't get it working:
https://forum.snapcraft.io/t/how-to-change-snapcraft-yaml-to-build-on-snapcraft-core18/12073/4

Current Wekan Snap version is based on devel branch, that has old Meteor 1.6.x version. I would like to delete devel branch, and move Wekan snap to use only master branch, that has newest Meteor 1.8.1. I have already deleted other old branches, because maintaining multiple branches takes huge amount of my time.

I would like to build Wekan for all Snap architectures:
  - amd64
  - arm64
  - armel
  - armhf
  - i386
  - powerpc
  - ppc64el
  - s390x

And if possible, more architectures if there is Node+MongoDB+NPM for it.

Please help by adding pull request to master branch, to make it working. I have tried to get Meteor 1.8.1 working with Snap without any success for many months.

**Comment by xet7** [comment]
I can't get newest Wekan working in Snap, I get these errors https://github.com/wekan/wekan/issues/2533#issuecomment-513506939 , can somebody help?
